Client/Coordinators/Router/DefaultRouter.swift

```diff
@@ -0,0 +1,99 @@
+
+    func present(_ viewController: UIViewController, animated: Bool = true, completion: (() -> Void)? = nil) {
+        if let completion = completion {
+            completions[viewController] = completion
+        }
+
+        viewController.presentationController?.delegate = self
+        navigationController.present(viewController, animated: animated, completion: nil)
+    }
+
+    func dismiss(animated: Bool = true, completion: (() -> Void)? = nil) {
+        navigationController.dismiss(animated: animated, completion: completion)
+    }
+
+    func push(_ viewController: UIViewController, animated: Bool = true, completion: (() -> Void)? = nil) {
+        // Avoid pushing UINavigationController onto stack
+        guard viewController is UINavigationController == false else {
+            return
+        }
+
+        if let completion = completion {
+            completions[viewController] = completion
+        }
+
+        navigationController.pushViewController(viewController, animated: animated)
+    }
+
+    func popViewController(animated: Bool = true) {
+        if let controller = navigationController.popViewController(animated: animated) {
+            runCompletion(for: controller)
+        }
+    }
+
+    func setRootViewController(_ viewController: UIViewController, hideBar: Bool = false) {
+        // Call all completions so all coordinators can be deallocated
+        completions.forEach { $0.value() }
+        navigationController.setViewControllers([viewController], animated: false)
+        navigationController.isNavigationBarHidden = hideBar
+    }
+
+    func popToRootViewController(animated: Bool = true) {
+        if let controllers = navigationController.popToRootViewController(animated: animated) {
+            controllers.forEach { runCompletion(for: $0) }
+        }
+    }
+
+    private func runCompletion(for controller: UIViewController) {
+        guard let completion = completions[controller] else { return }
+        completion()
+        completions.removeValue(forKey: controller)
+    }
+
+    // MARK: - UINavigationControllerDelegate
+
+    func navigationController(_ navigationController: UINavigationController,
+                              didShow viewController: UIViewController,
+                              animated: Bool) {
+        checkNavigationCompletion(for: navigationController)
+    }
+
+    func checkNavigationCompletion(for navigationController: NavigationController) {
+        // Read the view controller we’re moving from, then check whether our view controller array already contains
+        // that view controller. If it does it means we’re pushing a different view controller on top rather than
+        // popping it, so exit. Otherwise run the completion of that popped view controller.
+        guard let fromViewController = navigationController.fromViewController,
+              !navigationController.viewControllers.contains(fromViewController) else {
+            return
+        }
```

Comment:
GUARD POLICE!
```suggestion
        guard let fromViewController = navigationController.fromViewController,
              !navigationController.viewControllers.contains(fromViewController)
        else { return }
```

I really need to write a custom swiftlint rule for this.